File: hasura/examples/pokemon.js
/*
This is an example snippet - you should consider tailoring it
to your service.
*/
/*
Add these to your `package.json`:
  "node-fetch": "^2.5.0"
*/

// Node doesn't implement fetch so we have to import it
const fetch =require("node-fetch");

async function fetchGraphQL(operationsDoc, operationName, variables) {
  const result = await fetch(
    "http://localhost:80/graphql/v1beta",
    {
      method: "POST",
      body: JSON.stringify({
        query: operationsDoc,
        variables: variables,
        operationName: operationName
      })
    }
  );

  return await result.json();
}

const operationsDoc = `
  query pokemon_details {
    species: pokemon_v2_pokemonspecies(where: {name: {_eq: "staryu"}}) {
      name
      base_happiness
      is_legendary
      is_mythical
      generation: pokemon_v2_generation {
        name
      }
      habitat: pokemon_v2_pokemonhabitat {
        name
      }
      pokemon: pokemon_v2_pokemons_aggregate(limit: 1) {
        nodes {
          height
          name
          id
          weight
          abilities: pokemon_v2_pokemonabilities_aggregate {
            nodes {
              ability: pokemon_v2_ability {
                name
              }
            }
          }
          stats: pokemon_v2_pokemonstats {
            base_stat
            stat: pokemon_v2_stat {
              name
            }
          }
          types: pokemon_v2_pokemontypes {
            slot
            type: pokemon_v2_type {
              name
            }
          }
          levelUpMoves: pokemon_v2_pokemonmoves_aggregate(where: {pokemon_v2_movelearnmethod: {name: {_eq: "level-up"}}}, distinct_on: move_id) {
            nodes {
              move: pokemon_v2_move {
                name
              }
              level
            }
          }
          pokemon_v2_encounters_aggregate {
            aggregate {
              count
            }
          }
          pokemon_v2_pokemonitems(where: {pokemon_v2_version: {name: {_eq: "firered"}}}) {
            pokemon_v2_item {
              name
              cost
            }
            rarity
          }
        }
      }
      flavorText: pokemon_v2_pokemonspeciesflavortexts(where: {pokemon_v2_language: {name: {_eq: "en"}}, pokemon_v2_version: {name: {_eq: "firered"}}}) {
        flavor_text
      }
    }
  }
`;

function fetchPokemon_details() {
  return fetchGraphQL(
    operationsDoc,
    "pokemon_details",
    {}
  );
}

async function startFetchPokemon_details() {
  const { errors, data } = await fetchPokemon_details();

  if (errors) {
    // handle those errors like a pro
    console.error(errors);
  }

  // do something great with this precious data
  console.log(data);
}

startFetchPokemon_details();

File: graphql/examples/go/pokemon.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
)

type Operation struct {
	Query         string                 `json:"query"`
	Variables     map[string]interface{} `json:"variables"`
	OperationName string                 `json:"operationName"`
}

var (
	pokemonDetails = Operation{
		OperationName: "pokemon_details",
		Variables:     map[string]interface{}{
			"name": "staryu",
		},
		Query: `
query pokemon_details($name: String) {
	species: pokemonspecies(where: {name: {_eq: $name}}) {
	name
	base_happiness
	is_legendary
	is_mythical
	generation: generation {
		name
	}
	habitat: pokemonhabitat {
		name
	}
	pokemon: pokemons_aggregate(limit: 1) {
		nodes {
		height
		name
		id
		weight
		abilities: pokemonabilities_aggregate {
			nodes {
			ability: ability {
				name
			}
			}
		}
		stats: pokemonstats {
			base_stat
			stat: stat {
			name
			}
		}
		types: pokemontypes {
			slot
			type: type {
			name
			}
		}
		levelUpMoves: pokemonmoves_aggregate(where: {movelearnmethod: {name: {_eq: "level-up"}}}, distinct_on: move_id) {
			nodes {
			move: move {
				name
			}
			level
			}
		}
		foundInAsManyPlaces: encounters_aggregate {
			aggregate {
			count
			}
		}
		fireRedItems: pokemonitems(where: {version: {name: {_eq: "firered"}}}) {
			item {
			name
			cost
			}
			rarity
		}
		}
	}
	flavorText: pokemonspeciesflavortexts(where: {language: {name: {_eq: "en"}}, version: {name: {_eq: "firered"}}}) {
		flavor_text
	}
	}
}
	  `,
	}
)

func main() {
	url := "https://beta.pokeapi.co/graphql/v1beta"
	body, err := json.Marshal(pokemonDetails)
	if err != nil {
		log.Fatal(err)
	}

	resp, err := http.Post(url, "", bytes.NewReader(body))
	if err != nil {
		log.Fatal(err)
	}
	defer resp.Body.Close()
	body, err = ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(string(body))
}
